src/cargo/core/compiler/mod.rs

```diff
@@ -726,6 +726,16 @@ fn prepare_rustc(
     if unit.target.is_test() || unit.target.is_bench() {
         let tmp = cx.files().layout(unit.kind).prepare_tmp()?;
         base.env("CARGO_TARGET_TMPDIR", tmp.display().to_string());
+
+        let cargo_manifest_dir = unit.pkg.root();
+        let cargo_workspace_dir = cx.bcx.ws.root();
+        let cargo_workspace_dir_rel_path =
+            pathdiff::diff_paths(cargo_workspace_dir, cargo_manifest_dir)
+                .expect("both paths are absolute");
```

Comment:
Can you add a test where this is being used in a published package?

How that would work
- Publish a package with a test using the variable
- Create a package that depends on the published package
- `cargo test -p <published-package-name>` (since cargo lets you test dependencies)

For the published package, `CARGO_WORKSPACE_DIR` should be the same as `CARGO_MANIFEST_DIR`.

From https://github.com/rust-lang/cargo/issues/3946#issuecomment-1402464796:

> There was the question of whether this is for the "workspace" of what is being built now or the final artifact. We felt that it would best be associated with what is being built right now, either exposing macros for dependents to use or have them pass paths down into what is being compiled. That has clearer semantics and doesn't run into inverted dependencies where the crate knows what it is being compiled fir. This means that for the registry crates, the path would end up being the same as CARGO_MANIFEST_DIR